GitCommands/RevisionReader.cs

```diff
@@ -42,18 +42,107 @@ public sealed class RevisionReader
         private readonly Encoding _logOutputEncoding;
         private readonly long _sixMonths;
 
-        public RevisionReader(GitModule module)
-            : this(module, module.LogOutputEncoding, new DateTimeOffset(DateTime.Now.ToUniversalTime() - TimeSpan.FromDays(30 * 6)).ToUnixTimeSeconds())
+        // reflog selector to identify stashes
+        private readonly bool _hasReflogSelector;
+
+        private string LogFormat => _hasReflogSelector ? FullFormat.Replace("%B", "%gD%n%B") : FullFormat;
+
+        public RevisionReader(GitModule module, bool hasReflogSelector)
+            : this(module, hasReflogSelector, module.LogOutputEncoding, new DateTimeOffset(DateTime.Now.ToUniversalTime() - TimeSpan.FromDays(30 * 6)).ToUnixTimeSeconds())
         {
         }
 
-        internal RevisionReader(GitModule module, Encoding logOutputEncoding, long sixMonths)
+        private RevisionReader(GitModule module, bool hasReflogSelector, Encoding logOutputEncoding, long sixMonths)
         {
             _module = module;
+            _hasReflogSelector = hasReflogSelector;
             _logOutputEncoding = logOutputEncoding;
             _sixMonths = sixMonths;
         }
 
+        /// <summary>
+        /// Get the git-stash GitRevisions
+        /// </summary>
+        /// <param name="token">Cancellation token</param>
+        /// <returns>List with GitRevisions</returns>
+        public List<GitRevision> GetStashes(CancellationToken token)
+        {
+            Debug.Assert(_hasReflogSelector, "_hasReflogSelector must be set to get the reflog selectors (to identify stashes)");
+            GitArgumentBuilder arguments = new("stash")
+            {
+                "list",
+                "-z",
+                $"--pretty=format:\"{LogFormat}\""
+            };
+
+            List<GitRevision> stashes = new();
+            using (IProcess process = _module.GitCommandRunner.RunDetached(arguments, redirectOutput: true, outputEncoding: GitModule.LosslessEncoding))
+            {
+                byte[] buffer = new byte[4096];
+
+                foreach (ArraySegment<byte> chunk in process.StandardOutput.BaseStream.ReadNullTerminatedChunks(ref buffer))
+                {
+                    token.ThrowIfCancellationRequested();
+
+                    if (TryParseRevision(chunk, out GitRevision? revision))
+                    {
+                        stashes.Add(revision);
+                    }
+                }
+            }
+
+            return stashes;
+        }
+
+        /// <summary>
+        /// Get the GitRevisions for the listed files,
+        /// excluding commits without changes.
+        /// </summary>
+        /// <param name="untracked">commit list</param>
+        /// <param name="token">Cancellation token</param>
+        /// <returns>List with GitRevisions</returns>
+        public List<GitRevision> GetRevisionsFromList(IList<ObjectId> untracked, CancellationToken token)
```

Comment:
What is the benefit with this?
I find it easier to read checking the size (and slightly more efficient even if that is not a focus here).

Reply:
We constantly reading revisions, and this code is on the hot path. So we want to avoid memory churn as much as possible.

Reply:
I did not see any improvement for allocation in your proposal, just a slower check.

For performance and allocation when loading, the important path is the reading of revisions (in RevisionReader that is optimized with my best effort) and in RevGrid (can likely be a few tuning done). The general improvement is to avoid calling Git when not needed.
I am adding an if and a null allocation in this PR to the rev reading loop, which will add ms.
(Replacing String.Empty with "" and removing precompiled regex is a more obvious optimization...)

Reply:
I made a simple refactoring and changed this code slightly.